src/Sylius/Component/Core/Payment/Provider/OrderPaymentProvider.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Payment\Provider;
+
+use SM\Factory\FactoryInterface as StateMachineFactoryInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+use Sylius\Component\Payment\Exception\UnresolvedDefaultPaymentMethodException;
+use Sylius\Component\Payment\Factory\PaymentFactoryInterface;
+use Sylius\Component\Payment\PaymentTransitions;
+use Sylius\Component\Payment\Resolver\DefaultPaymentMethodResolverInterface;
+use Sylius\Component\Resource\StateMachine\StateMachineInterface;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+final class OrderPaymentProvider implements OrderPaymentProviderInterface
+{
+    /**
+     * @var DefaultPaymentMethodResolverInterface
+     */
+    private $defaultPaymentMethodResolver;
+
+    /**
+     * @var PaymentFactoryInterface
+     */
+    private $paymentFactory;
+
+    /**
+     * @var StateMachineFactoryInterface
+     */
+    private $stateMachineFactory;
+
+    /**
+     * @param DefaultPaymentMethodResolverInterface $defaultPaymentMethodResolver
+     * @param PaymentFactoryInterface $paymentFactory
+     * @param StateMachineFactoryInterface $stateMachineFactory
+     */
+    public function __construct(
+        DefaultPaymentMethodResolverInterface $defaultPaymentMethodResolver,
+        PaymentFactoryInterface $paymentFactory,
+        StateMachineFactoryInterface $stateMachineFactory
+    ) {
+        $this->defaultPaymentMethodResolver = $defaultPaymentMethodResolver;
+        $this->paymentFactory = $paymentFactory;
+        $this->stateMachineFactory = $stateMachineFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function provideOrderPayment(OrderInterface $order, $targetState)
+    {
+        /** @var $payment PaymentInterface */
+        $payment = $this->paymentFactory->createWithAmountAndCurrencyCode($order->getTotal(), $order->getCurrencyCode());
+
+        $paymentMethod = $this->getDefaultPaymentMethod($payment, $order);
+        $lastPayment = $this->getLastPayment($order);
+
+        if (null !== $lastPayment) {
+            $paymentMethod = $lastPayment->getMethod();
+        }
+
+        if (null === $paymentMethod) {
+            return null;
```

Comment:
We should throw an exception, if I run `$provider->provideOrderPayment(...)` then I do not expect `null`.